ec2/utils.py

```diff
@@ -13,61 +13,200 @@
+    """
 
-def validate_state(instance, state, timeout_length, check_interval, ctx):
-    """ Check if an EC2 instance is in a particular state.
-    :param instance: And EC2 instance.
-    :param state: The state code (pending = 0, running = 16,
-                  shutting down = 32, terminated = 48, stopping = 64,
-                  stopped = 80
-    :param timeout_length: How long to wait for a positive answer
-           before we stop checking.
-    :param check_interval: How long to wait between checks.
-    :return: bool (True the desired state was reached, False, it was not.)
+    if key not in ctx.node.properties.keys() \
+            or ctx.node.properties.get(key) is None:
+        raise NonRecoverableError('{0} is a required input .'
+                                  'Unable to create.'.format(key))
+
+
+def save_key_pair(key_pair_object, ctx):
+    """ Saves the key pair to the file specified in the blueprint. """
+
+    ctx.logger.debug('Attempting to save the key_pair_object.')
+
+    try:
+        key_pair_object.save(ctx.node.properties['private_key_path'])
+    except (boto.exception.BotoClientError, OSError) as e:
+        raise NonRecoverableError('Unable to save key pair to file: {0}.'
+                                  'OS Returned: {1}'.format(
+                                      ctx.node.properties['private_key_path'],
+                                      str(e)))
+
+    path = os.path.expanduser(ctx.node.properties['private_key_path'])
+    key_path = os.path.join(path,
+                            '{0}{1}'.format(
+                                ctx.node.properties['resource_id'],
+                                '.pem'))
+
+    os.chmod(key_path, 0600)
+
+
+def delete_key_pair(key_pair_name, ctx):
+    """ Deletes the key pair in the file specified in the blueprint. """
+
+    ctx.logger.debug('Attempting to save the key_pair_object.')
+
+    path = os.path.expanduser(ctx.node.properties['private_key_path'])
+    file = os.path.join(path,
+                        '{0}{1}'.format(
+                            ctx.node.properties['resource_id'],
+                            '.pem'))
+    if os.path.exists(file):
+        try:
+            os.remove(file)
+        except OSError:
+            raise NonRecoverableError('Unable to save key pair to file: {0}.'
+                                      'OS Returned: {1}'.format(path,
+                                                                str(OSError)))
+
+
+def search_for_key_file(ctx):
+    """ Indicates whether the file exists locally. """
+
+    path = os.path.expanduser(ctx.node.properties['private_key_path'])
+    file = os.path.join(path,
+                        '{0}{1}'.format(
+                            ctx.node.properties['resource_id'],
+                            '.pem'))
+    if os.path.exists(file):
+        return True
+    else:
+        return False
+
+
+def get_instance_attribute(attribute, check_interval, ctx):
```

Comment:
`get_instance_attribute` should be a generic function for retrieving an attribute for an instance OBJECT, regardless of its state. We should check if the caller requires the instance to be running.. or so.